### PR TITLE
Mutox classifier

### PR DESCRIPTION
**Pull request by David-OC17**
## Elements to include in PR

- [ ] Migrate Mutox code from [Seamless Communication]([src/seamless_communication/toxicity/mutox](https://github.com/facebookresearch/seamless_communication/tree/main/src/seamless_communication/toxicity/mutox))
- [ ] Add unit tests for the pipeline
- [ ] (Optional) add softmax computation on top of classifier

**Note:** A follow-up PR will be needed in the Seamless Communication repository to account for the migration being made here.

### CODE REVIEW

File: examples/mutox_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sonar.inference_pipelines.speech import SpeechInferenceParams\n",
+    "from mutox.speech_pipeline import MutoxSpeechClassifierPipeline\n",
+    "\n",
+    "pipeline_builder = MutoxSpeechClassifierPipeline.load_model_from_name(\n",
+    "    mutox_classifier_name =\"mutox\",\n",
```
Comment (David-OC17) [inline examples/mutox_example.ipynb:94]: Unable to import at this time*

File: mutox/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
```
Comment (avidale) [inline mutox/__init__.py:1]: I would prefer adding the `mutox` folder within the `sonar/models` directory, next to other models: `blaser`, `sonar_text`, etc.

File: mutox/cli/README.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
+# MuTox: MuTox: Universal MUltilingual Audio-based TOXicity Dataset and Zero-shot Detector
+
+MuTox, the first highly multilingual audio-based dataset with toxicity labels.
```
Comment (avidale) [inline mutox/cli/README.md:3]: And instead of porting this long readme, I would add a small section in the root readme file of this repo, that would point to the full readme in the CS repo. This is because the Mutox paper is already immutable, and it points to the CS repo, so it makes sense to me to keep the detailed Mutox readme there.

@alexmourachko @artemru what do you think?

File: tests/unit_tests/test_mutox.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+from sonar.models.mutox.builder import MutoxConfig, MutoxClassifierBuilder, create_mutox_model
+from sonar.models.mutox.classifier import MutoxClassifier
+from sonar.models.mutox.loader import (
+    convert_mutox_checkpoint,
+)
+
+# Builder tests
+
+@pytest.mark.parametrize("input_size", [256, 512, 1024])
+@pytest.mark.parametrize("device", [torch.device("cpu")])
```
Comment (artemru) [inline tests/unit_tests/test_mutox.py:21]: there's CPU object in fs2

File: sonar/models/mutox/classifier.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+class MutoxClassifier(nn.Module):
+    def __init__(
+        self,
+        model_all,
+    ):
+        super().__init__()
+        self.model_all = model_all
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        return self.model_all(inputs)
```
Comment (artemru) [inline sonar/models/mutox/classifier.py:26]: forward seems to be trivial here, it looks like model_all =  MutoxClassifier...
maybe we could move model layer creation from the builder to MutoxClassifier to make it more meaningful.

File: sonar/models/mutox/builder.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            nn.Linear(self.config.input_size, 512),
+        )
+
+        model_h2 = nn.Sequential(
+            nn.ReLU(),
+            nn.Linear(512, 128),
+        )
+
+        model_h3 = nn.Sequential(
+            nn.ReLU(),
```
Comment (artemru) [inline sonar/models/mutox/builder.py:57]: dont we want to parametrize the activation here ?

File: examples/mutox_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "/tmp/tmpqasvhgx6/commonvoice_example_en_clocks.wav\t-42.40079116821289\n",
```
Comment (avidale) [inline examples/mutox_example.ipynb:128]: These scores, -42 and -47, are not straightforward to interpret.
We should either mention in the top of the notebook that the model has been trained with a "Binary Cross Entropy loss with logits" objective (according) to the paper), and thus, to turn its output into a probability, we need to feed it to a sigmoid layer.

File: README.md [hunk 8]
```diff
@@ -142,6 +142,33 @@ print(blaser_qe(src=src_embs, mt=mt_embs).item())  # 4.708
...
+from sonar.inference_pipelines.text import TextToEmbeddingModelPipeline
+
+t2vec_model = TextToEmbeddingModelPipeline(
+    encoder="text_sonar_basic_encoder",
+    tokenizer="text_sonar_basic_encoder",
+    device=device,
+)
+text_column='lang_txt'
+classifier = load_mutox_model(
+    "mutox",
```
Comment (avidale) [inline README.md:160]: This code won't work unless we have a model card "mutox" in some place where fairseq2 model loader can access it. 
We have one such card [in the SC repo](https://github.com/facebookresearch/seamless_communication/blob/main/src/seamless_communication/cards/mutox.yaml), but we don't want the SONAR code to be dependent on it. 
So maybe we should copy the card to `sonar/cards`. And we might need to change the name of the new card; otherwise, there will be a model name onflict for those who have both SONAR and seamless_communication packages installed. 
@artemru how would you suggest to handle this gracefully?

Comment (David-OC17) [reply README.md:160]: For now, I just added the card to `sonar/cards` but it might be best if it could be accessed through `https://dl.fbaipublicfiles.com/SONAR/`, but I do not have access to add it there. I guess another solution not to pull it from  `https://dl.fbaipublicfiles.com/seamless/models/mutox.pt` would be to add a release to this GitHub repo and add a section to download the model from there.

You tell me what's best.

File: README.md [hunk 9]
```diff
@@ -142,6 +142,33 @@ print(blaser_qe(src=src_embs, mt=mt_embs).item())  # 4.708
...
+[MuTox](https://github.com/facebookresearch/seamless_communication/tree/main/src/seamless_communication/cli/toxicity/mutox), the first highly multilingual audio-based dataset with toxicity labels. The dataset consists of 20k audio utterances for English and Spanish, and 4k for the other 19 languages, and uses the multi-model and multilingual encoders from SONAR.
+
+```Python
+from sonar.models.mutox.loader import load_mutox_model
+from sonar.inference_pipelines.text import TextToEmbeddingModelPipeline
+
+t2vec_model = TextToEmbeddingModelPipeline(
+    encoder="text_sonar_basic_encoder",
+    tokenizer="text_sonar_basic_encoder",
+    device=device,
```
[the remark this answers is not in the document]
Comment (David-OC17) [reply README.md:156]: I don't get how to include that during the creation/calling of the encoder...

[the remark this answers is not in the document]
Comment (avidale) [reply README.md:164]: Never mind. I confused `TextToEmbeddingModelPipeline` with another constructor which allows to explicitly specify the model dtype.

File: README.md [hunk 10]
```diff
@@ -142,6 +142,49 @@ print(blaser_qe(src=src_embs, mt=mt_embs).item())  # 4.708
...
+    dtype=dtype,
+).eval()
+
+with torch.inference_mode():
+    emb = t2vec_model.predict(["De peur que le pays ne se prostitue et ne se remplisse de crimes."], source_lang='fra_Latn')
+    x = classifier(emb.to(device).to(dtype)) # tensor([[-19.7812]], device='cuda:0', dtype=torch.float16)
+
+with torch.inference_mode():
+    emb = t2vec_model.predict(["She worked hard and made a significant contribution to the team."], source_lang='fra_Latn')
```
Comment (avidale) [inline README.md:178]: `source_lang='eng_Latn'`?

File: examples/mutox_example.ipynb (hunk 7 above)
Comment (avidale) [reply examples/mutox_example.ipynb:135]: The same thing: you resolved the comment, but the notebook still doesn't feature a code change or an explanation of what kind of output the current code generats.

File: sonar/inference_pipelines/mutox_speech.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+class MutoxSpeechClassifierPipeline(SpeechToEmbeddingPipeline):
+    def __init__(
+        self,
+        mutox_classifier: Union[str, MutoxClassifier],
+        encoder: Union[str, SonarEncoderModel],
+        device: Device = CPU_DEVICE,
+    ) -> None:
+        super().__init__(encoder)
+        self.model.to(device).eval()
+        self.mutox_classifier = mutox_classifier.to(device).eval()
```
[the remark this answers is not in the document]
Comment (David-OC17) [reply sonar/inference_pipelines/mutox_speech.py:40]: done by loading if str

File: examples/mutox_example.ipynb (hunk 7 above)
Comment (David-OC17) [reply examples/mutox_example.ipynb:135]: Added simple comment specifying the training & output

File: sonar/models/mutox/classifier.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+@dataclass
+class MutoxConfig:
+    """Holds the configuration of a Mutox Classifier model."""
+
+    # size of the input embedding supported by this model
+    input_size: int
+
+    # add sigmoid as last layer to output probability
+    output_prob: bool = False
```
[the remark this answers is not in the document]
Comment (David-OC17) [reply sonar/models/mutox/classifier.py:36]: Adding sigmoid layer if optional parameter in forward() is true

File: sonar/models/mutox/classifier.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        model_all,
+    ):
+        super().__init__()
+        self.model_all = model_all
+
+    def forward(self, inputs: torch.Tensor, output_prob: bool = False) -> torch.Tensor:
+        if output_prob:
+            self.model_all.add_module("sigmoid", nn.Sigmoid())
+        else:
+            self.model_all.add_module("linear", nn.Linear(128, 1))
```
Comment (avidale) [inline sonar/models/mutox/classifier.py:28]: You are doing something very weird here: changing the model structure during its inference. 
This is incorrect for several reasons:
1. The model needs to have the `Linear(128, 1)` layer regardless of whether we want its output to be a logit or a probability, so you didn't need to remove it from `MutoxClassifierBuilder.build_model`. 
2. When a `Linear(128, 1)` (or any other layer with trainable parameters) is added, its parameters are initialized randomly. When a model is loaded with a `ModelLoader`, it overrides these random parameters with the values loaded from a disk. But if you add the model after this, it stays random. 
3. If you run `forward` several times in a row, the sigmoid or linear layers would be added every time.

So I suggest you do the following:
1. Add the last linear layer back in `MutoxClassifierBuilder.build_model`.
2. In `MutoxClassifier.forward`, you conditionally apply `torch.sigmoid` to the model outputs.

File: tests/unit_tests/test_mutox.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+    # Verify only 'model_all.' keys are retained in the converted dictionary
+    assert "model" in converted, "Converted checkpoint should contain a 'model' key"
+    assert (
+        "model_all.layer1.weight" in converted["model"]
+    ), "Expected 'model_all.layer1.weight'"
+    assert (
+        "model_all.layer1.bias" in converted["model"]
+    ), "Expected 'model_all.layer1.bias'"
+    assert "non_model_key" not in converted["model"], "Unexpected 'non_model_key'"
```
Comment (avidale) [inline tests/unit_tests/test_mutox.py:105]: 1. I think we also need a unit test for the `output_prob` argument. 
2. Please add an integration test that compares the outputs of the `sonar_mutox` classifier  for a few given texts with their expected values (e.g. the inputs and outputs that you already demonstrate in the readme). Just to make sure that we don't break the model in some unexpected way which is not detectable with unit tests that check only the output shape.

File: tests/unit_tests/test_mutox.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    """Test that MutoxClassifier forward pass returns expected output shape."""
+    test_model = nn.Sequential(
+        nn.Linear(10, 5),
+        nn.ReLU(),
+        nn.Linear(5, 1),
+    )
+    model = MutoxClassifier(test_model)
+
+    test_input = torch.randn(3, 10)
+    output = model(test_input)
```
Comment (avidale) [inline tests/unit_tests/test_mutox.py:68]: Actually, this test is working! It is breaking right now, because the at some point, the matrix shapes are not compatible (as a result of messing with the last linear layer). 

I am not sure if you are paying attention to the failing tests in the PR interface, but they are actually informative and useful:
image

Comment (David-OC17) [reply tests/unit_tests/test_mutox.py:68]: I have been reading them, I resolved most of the problems related to mypy that appeared after the first migration of the code

File: tests/unit_tests/test_mutox.py (hunk 14 above)
Comment (David-OC17) [reply tests/unit_tests/test_mutox.py:127]: I added integration tests for checking the output of both the model with linear layer output, and one for the sigmoid one. Now, the thing is that the output of the sigmoid, given the range that the model usually gives (high positive or negative numbers), is giving 0 or 1 for most input sentences. Should we add a "temperature" parameter of some sort? to try to make the values more informative

Comment (avidale) [reply tests/unit_tests/test_mutox.py:127]: No, I don't think temperature would make sense here, keep it as is.

**Comment by David-OC17** [comment]
Got it down to that one mypy error, cannot solve it still.

**Comment by avidale** [comment]
> Got it down to that one mypy error, cannot solve it still.

If you absolutely cannot, I think, it is fine to tell Mypy just to ignore this like (by adding a comment `# type: ignore` on the corresponding line.
